doc/Support/Configuration.md
# Configuration Docs

KartsNMS configuration is a set of key values.

The config is stored in two places:
Database: This applies to all pollers and can be set with either `lnms config:set` or in the Web UI. Database config takes precedence over config.php.
config.php: This applies to the local poller only.  Configs set here will be disabled in the Web UI to prevent unexpected behaviour.

The KartsNMS uses dot notation for config items:

| Database | config.php |
| -------- | ---------- |
| `snmp.community` | `$config['snmp']['community']` |
| `snmp.community.+` | `$config['snmp']['community'][]` |
| `snmp.v3.0.authalgo` | `$config['snmp']['v3'][0]['authalgo']` |

> The documentation has not been updated to reflect using `lnms config:set` to
> set config items, but it will work for all settings.  Not all settings have
> been defined in KartsNMS, but they can still be set with the `--ignore-checks`
> option.  Without that option input is checked for correctness, that does not
> mean it is not possible to set bad values.  Please report missing settings.

## CLI
`lnms config:get` will fetch the current config settings (composite of database, config.php, and defaults).  
`lnms config:set` will set the config setting in the database.  Calling `lnms config:set` on a setting with no value will reset it to the default value.

If you set up bash completion, you can use tab completion to find config settings.

### Getting a list of all current values

To get a complete list of all the current values, you can use the command `lnms config:get --dump`. The output may not be desirable, so you can use the `jq` package to pretty print it. Then it would be `lnms config:get --dump | jq`.

Example output:
```
kartsnms@kartsnms:~$ lnms config:get --dump | jq 
{
  "install_dir": "/opt/kartsnms",
  "active_directory": {
    "users_purge": 0
  },
  "addhost_alwayscheckip": false,
  "alert": {
    "ack_until_clear": false,
    "admins": true,
    "default_copy": true,
    "default_if_none": false,
    "default_mail": false,
    "default_only": true,
    "disable": false,
    "fixed-contacts": true,
    "globals": true,
    "syscontact": true,
    "transports": {
      "mail": 5
    },
    "tolerance_window": 5,
    "users": false,
    ...
```

### Examples

```bash
lnms config:get snmp.community
  [
      "public"
  ]


lnms config:set snmp.community.+ testing

lnms config:get snmp.community
  [
      "public",
      "testing"
  ]


lnms config:set snmp.community.0 private

lnms config:get snmp.community
  [
      "private",
      "testing"
  ]

lnms config:set snmp.community test
  Invalid format

lnms config:set snmp.community '["test", "othercommunity"]'

lnms config:get snmp.community
  [
      "test",
      "othercommunity"
  ]

lnms config:set snmp.community

  Reset snmp.community to the default? (yes/no) [no]:
  > yes


lnms config:get snmp.community
  [
      "public"
  ]
```

## Pre-load configuration

This feature is primarily for docker images and other automation.
When installing KartsNMS for the first time with a new database you can place yaml key value files
in `database/seeders/config` to pre-populate the config database.

Example snmp.yaml
```yaml
snmp.community:
    - public
    - private
snmp.max_repeaters: 30
```

## Directories

```bash
lnms config:set temp_dir /tmp
```

The temporary directory is where images and other temporary files are
created on your filesystem.

```bash
lnms config:set log_dir /opt/kartsnms/logs
```

Log files created by KartsNMS will be stored within this directory.

## Database config

Set these variables either in .env (/opt/kartsnms/.env by default) or in the environment.

```dotenv
DB_HOST=127.0.0.1
DB_DATABASE=kartsnms
DB_USERNAME=DBUSER
DB_PASSWORD="DBPASS"
```

Use non-standard port:

```dotenv
DB_PORT=3306
```

Use a unix socket:

```dotenv
DB_SOCKET=/run/mysqld/mysqld.sock
```

## Core

### PHP Settings

You can change the memory limits for php within `config.php`. The
value is in Megabytes and should just be an int value:

`lnms config:set php_memory_limit 128`

### Programs

A lot of these are self explanatory so no further information may be
provided. Any extensions that have dedicated  documentation page will
be linked to rather than having the config provided.

#### RRDTool

> You can configure these options within the WebUI now, please avoid
> setting these options within config.php
>
> Settings -> External Settings -> RRDTool Setup

!!! setting "external/binaries"
    ```bash
    lnms config:set rrdtool /usr/bin/rrdtool
    ```

Please see [1 Minute polling](1-Minute-Polling.md) for information on
configuring your install to record data more frequently.

#### fping

!!! setting "external/binaries"
    ```bash
    lnms config:set fping /usr/bin/fping
    lnms config:set fping6 fping6
    ```

!!! setting "poller/ping"
    ```bash
    lnms config:set fping_options.timeout 500
    lnms config:set fping_options.count 3
    lnms config:set fping_options.interval 500
    lnms config:set fping_options.tos 184
    ```

`fping` configuration options:

* `timeout` (`fping` parameter `-t`): Amount of time that fping waits
  for a response to its first request (in milliseconds). **See note
  below**
* `count` (`fping` parameter `-c`): Number of request packets to send
  to each target.
* `interval` (`fping` parameter `-p`): Time in milliseconds that fping
  waits between successive packets to an individual target.
* `tos` (`fping`parameter `-O`): Set the type of service flag (TOS). Value can be either decimal or hexadecimal (0xh) format. Can be used to ensure that ping packets are queued in following QOS mecanisms in the network. Table is accessible in the [TOS Wikipedia page](https://en.wikipedia.org/wiki/Type_of_service).

> NOTE: Setting a higher timeout value than the interval value can
> lead to slowing down poller. Example:
>
> timeout: 3000
>
> count: 3
>
> interval: 500
>
> In this example, interval will be overwritten by the timeout value
> of 3000 which is 3 seconds. As we send three icmp packets (count:
> 3), each one is delayed by 3 seconds which will result in fping
> taking > 6 seconds to return results.

You can disable the fping / icmp check that is done for a device to be
determined to be up on a global or per device basis. **We don't advise
disabling the fping / icmp check unless you know the impact, at worst
if you have a large number of devices down then it's possible that the
poller would no longer complete in 5 minutes due to waiting for snmp
to timeout.**

Globally disable fping / icmp check:

```bash
lnms config:set icmp_check false
```

If you would like to do this on a per device basis then you can do so
under Device -> Edit -> Misc -> Disable ICMP Test? On

#### traceroute

KartsNMS uses traceroute to record debug information
when a device is down due to icmp AND you have
`lnms config:set debug.run_trace true` set.

!!! setting "external/binaries"
    ```bash
    lnms config:set traceroute /usr/bin/traceroute
    ```

#### SNMP

SNMP program locations.

!!! setting "external/binaries"
    ```bash
    lnms config:set snmpwalk /usr/bin/snmpwalk
    lnms config:set snmpget /usr/bin/snmpget
    lnms config:set snmpbulkwalk /usr/bin/snmpbulkwalk
    lnms config:set snmpgetnext /usr/bin/snmpgetnext
    lnms config:set snmptranslate /usr/bin/snmptranslate
    ```

#### Misc binaries
!!! setting "external/binaries"
```bash
lnms config:set whois /usr/bin/whois
lnms config:set ping /bin/ping
lnms config:set mtr /usr/bin/mtr
lnms config:set nmap /usr/bin/nmap
lnms config:set nagios_plugins /usr/lib/nagios/plugins
lnms config:set ipmitool /usr/bin/ipmitool
lnms config:set virsh /usr/bin/virsh
lnms config:set dot /usr/bin/dot
lnms config:set sfdp /usr/bin/sfdp
```

## Authentication

Generic Authentication settings.

Password minimum length for auth that allows user creation

```bash
lnms config:set password.min_length 8
```

## Proxy support

For alerting and the callback functionality, we support the use of a
http proxy setting. These can be any one of the following:

!!! setting "system/proxy"
    ```bash
    lnms config:set callback_proxy proxy.domain.com
    lnms config:set http_proxy proxy.domain.com
    ```

We can also make use of one of these environment variables which can be set in `/etc/environment`:

```bash
http_proxy=proxy.domain.com
https_proxy=proxy.domain.com
```

## RRDCached

Please refer to [RRDCached](../Extensions/RRDCached.md)

## WebUI Settings

```bash
lnms config:set base_url http://demo.kartsnms.org
```

KartsNMS will attempt to detect the URL you are using but you can override that here.

!!! setting "webui/style"
    ```bash
    lnms config:set site_style light
    ```

Currently we have a number of styles which can be set which will alter
the navigation bar look. dark, light and mono with light being the default.

```bash
lnms config:set webui.custom_css.+ css/custom/styles.css
```

You can override a large number of visual elements by creating your
own css stylesheet and referencing it here, place any custom css files
into  `html/css/custom` so they will be ignored by auto updates. You
can specify as many css files as you like, the order they are within
your config will be the order they are loaded in the browser.

!!! setting "webui/style"
    ```bash
    lnms config:set title_image images/custom/yourlogo.png
    ```

You can override the default logo with yours, place any custom images
files into `html/images/custom` so they will be ignored by auto updates.

```bash
lnms config:set page_refresh 300
```

Set how often pages are refreshed in seconds. The default is every 5
minutes. Some pages don't refresh at all by design.

```bash
lnms config:set front_page default
```

You can create your own front page by adding a blade file in `resources/views/overview/custom/`
and setting `front_page` to it's name.
For example, if you create `resources/views/overview/custom/foobar.blade.php`, set `front_page` to `foobar`.

!!! setting "webui/dashboard"
    ```bash
    lnms config:set webui.default_dashboard_id 0
    ```

Allows the specification of a global default dashboard page for any user who
has not set one in their user preferences.  Should be set to dashboard_id of an
existing dashboard that is shared or shared(read).  Otherwise, the system will
automatically create each user an empty dashboard called `Default` on their
first login.

```bash
lnms config:set login_message "Unauthorised access or use shall render the user liable to criminal and/or civil prosecution."
```

This is the default message on the login page displayed to users.

```bash
lnms config:set public_status true
```

If this is set to true then an overview will be shown on the login page of devices and the status.

```bash
lnms config:set show_locations true  # Enable Locations on menu
lnms config:set show_locations_dropdown true  # Enable Locations dropdown on menu
lnms config:set show_services false  # Disable Services on menu
lnms config:set int_customers true  # Enable Customer Port Parsing
lnms config:set summary_errors false  # Show Errored ports in summary boxes on the dashboard
lnms config:set customers_descr '["cust"]'  # The description to look for in ifDescr. Can have multiple '["cust","cid"]'
lnms config:set transit_descr '["transit"]'  # Add custom transit descriptions (array)
lnms config:set peering_descr '["peering"]'  # Add custom peering descriptions (array)
lnms config:set core_descr '["core"]'  # Add custom core descriptions  (array)
lnms config:set custom_descr '["This is Custom"]'  # Add custom interface descriptions (array)
lnms config:set int_transit true  # Enable Transit Types
lnms config:set int_peering true  # Enable Peering Types
lnms config:set int_core true  # Enable Core Port Types
lnms config:set int_l2tp false  # Disable L2TP Port Types
```

Enable / disable certain menus from being shown in the WebUI.

You are able to adjust the number and time frames of the quick select
time options for graphs and the mini graphs shown per row.

Quick select:

```bash
lnms config:set graphs.mini.normal '{
    "day": "24 Hours",
    "week": "One Week",
    "month": "One Month",
    "year": "One Year"
}'

lnms config:set graphs.mini.widescreen '{
    "sixhour": "6 Hours",
    "day": "24 Hours",
    "twoday": "48 Hours",
    "week": "One Week",
    "twoweek": "Two Weeks",
    "month": "One Month",
    "twomonth": "Two Months",
    "year": "One Year",
    "twoyear": "Two Years"
}'
```

Mini graphs:

```bash
lnms config:set graphs.row.normal '{
    "sixhour": "6 Hours",
    "day": "24 Hours",
    "twoday": "48 Hours",
    "week": "One Week",
    "twoweek": "Two Weeks",
    "month": "One Month",
    "twomonth": "Two Months",
    "year": "One Year",
    "twoyear": "Two Years"
}'
```

```bash
lnms config:set web_mouseover true
```

You can disable the mouseover popover for mini graphs by setting this to false.

```bash
lnms config:set enable_lazy_load true
```

You can disable image lazy loading by setting this to false.

```bash
lnms config:set overview_show_sysDescr true
```

Enable or disable the sysDescr output for a device.

!!! setting "webui/device"
    ```bash
    lnms config:set device_display_default '{{ $hostname }}'
    ```

This is a simple template to control the display of device names by default.
You can override this setting per-device.

You may enter any free-form text including one or more of the following template replacements:

| Template                    | Replacement                                                          |
|-----------------------------|----------------------------------------------------------------------|
| `{{ $hostname }}`           | The hostname or IP of the device that was set when added  *default   |
| `{{ $sysName_fallback }}`   | The hostname or sysName if hostname is an IP                         |
| `{{ $sysName }}`            | The SNMP sysName of the device, falls back to hostname/IP if missing |
| `{{ $ip }}`                 | The actual polled IP of the device, will not display a hostname      |

For example, `{{ $sysName_fallback }} ({{ $ip }})` will display something like `server (192.168.1.1)`

```bash
lnms config:set device_traffic_iftype.+ '/loopback/'
```

Interface types that aren't graphed in the WebUI. The default array
contains more items, please see misc/config_definitions.json for the full list.

```bash
lnms config:set enable_clear_discovery true
```

Administrators are able to clear the last discovered time of a device
which will force a full discovery run within the configured 5 minute cron window.

```bash
lnms config:set enable_footer true
```

Disable the footer of the WebUI by setting `enable_footer` to 0.

You can enable the old style network map (only available for
individual devices with links discovered via xDP) by setting:

```bash
lnms config:set gui.network-map.style old
```

```bash
lnms config:set percentile_value 90
```

Show the `X`th percentile in the graph instead of the default 95th percentile.

!!! setting "webui/graph"
    ```bash
    lnms config:set shorthost_target_length 15
    ```

The target maximum hostname length when applying the shorthost() function.
You can increase this if you want to try and fit more of the hostname in graph titles.
The default value is 12 However, this can possibly break graph
generation if this is very long.

You can enable dynamic graphs within the WebUI under Global Settings
-> Webui Settings -> Graph Settings.

Graphs will be movable/scalable without reloading the page:
![Example dynamic graph usage](img/dynamic-graph-usage.gif)

## Stacked Graphs

You can enable stacked graphs instead of the default inverted
graphs. Enabling them is possible via webui Global Settings -> Webui
Settings -> Graph settings -> Use stacked graphs

## Add host settings

The following setting controls how hosts are added.  If a host is
added as an ip address it is checked to ensure the ip is not already
present. If the ip is present the host is not added. If host is added
by hostname this check is not performed.  If the setting is true
hostnames are resolved and the check is also performed.  This helps
prevents accidental duplicate hosts.

```bash
lnms config:set addhost_alwayscheckip false # true - check for duplicate ips even when adding host by name.

CHANGELOG.md
The changelog can be found at our [docs](https://docs.kartsnms.org/#General/Changelog/) site.


KartsNMS/Config.php
<?php
/**
 * Config.php
 *
 * Config convenience class to access and set config variables.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * @link       https://www.itkarts.com
 *
 * @copyright  2017 Tony Murray
 * @author     Tony Murray <murraytony@gmail.com>
 */

namespace KartsNMS;

use App\Models\Callback;
use App\Models\GraphType;
use Exception;
use Illuminate\Database\QueryException;
use Illuminate\Support\Arr;
use Illuminate\Support\Str;
use KartsNMS\Data\Store\Rrd;
use KartsNMS\Util\Debug;
use KartsNMS\Util\Version;
use Log;

class Config
{
    private static $config;

    /**
     * Load the config, if the database connected, pull in database settings.
     *
     * return &array
     */
    public static function load()
    {
        // don't reload the config if it is already loaded, reload() should be used for that
        if (self::isLoaded()) {
            return self::$config;
        }

        // merge all config sources together config_definitions.json > db config > config.php
        self::loadDefaults();
        self::loadDB();
        self::loadUserConfigFile(self::$config);

        // final cleanups and validations
        self::processConfig();

        // set to global for legacy/external things (is this needed?)
        global $config;
        $config = self::$config;

        return self::$config;
    }

    /**
     * Reload the config from files/db
     *
     * @return mixed
     */
    public static function reload()
    {
        self::$config = null;

        return self::load();
    }

    /**
     * Get the config setting definitions
     *
     * @return array
     */
    public static function getDefinitions()
    {
        return json_decode(file_get_contents(base_path('misc/config_definitions.json')), true)['config'];
    }

    private static function loadDefaults()
    {
        self::$config['install_dir'] = base_path();
        $definitions = self::getDefinitions();

        foreach ($definitions as $path => $def) {
            if (array_key_exists('default', $def)) {
                Arr::set(self::$config, $path, $def['default']);
            }
        }

        // load macros from json
        $macros = json_decode(file_get_contents(base_path('misc/macros.json')), true);
        Arr::set(self::$config, 'alert.macros.rule', $macros);

        self::processDefaults();
    }

    /**
     * Load the user config from config.php
     *
     * @param  array  $config  (this should be self::$config)
     */
    private static function loadUserConfigFile(&$config)
    {
        // Load user config file
        if (is_file(base_path('config.php'))) {
            @include base_path('config.php');
        }
    }

    /**
     * Get a config value, if non existent null (or default if set) will be returned
     *
     * @param  string  $key  period separated config variable name
     * @param  mixed  $default  optional value to return if the setting is not set
     * @return mixed
     */
    public static function get($key, $default = null)
    {
        if (isset(self::$config[$key])) {
            return self::$config[$key];
        }

        if (! Str::contains($key, '.')) {
            return $default;
        }

        return Arr::get(self::$config, $key, $default);
    }

    /**
     * Unset a config setting
     * or multiple
     *
     * @param  string|array  $key
     */
    public static function forget($key)
    {
        Arr::forget(self::$config, $key);
    }

    /**
     * Get a setting from a device, if that is not set,
     * fall back to the global config setting prefixed by $global_prefix
     * The key must be the same for the global setting and the device setting.
     *
     * @param  array  $device  Device array
     * @param  string  $key  Name of setting to fetch
     * @param  string  $global_prefix  specify where the global setting lives in the global config
     * @param  mixed  $default  will be returned if the setting is not set on the device or globally
     * @return mixed
     */
    public static function getDeviceSetting($device, $key, $global_prefix = null, $default = null)
    {
        if (isset($device[$key])) {
            return $device[$key];
        }

        if (isset($global_prefix)) {
            $key = "$global_prefix.$key";
        }

        return self::get($key, $default);
    }

    /**
     * Get a setting from the $config['os'] array using the os of the given device
     *
     * @param  string  $os  The os name
     * @param  string  $key  period separated config variable name
     * @param  mixed  $default  optional value to return if the setting is not set
     * @return mixed
     */
    public static function getOsSetting($os, $key, $default = null)
    {
        if ($os) {
            \KartsNMS\Util\OS::loadDefinition($os);

            if (isset(self::$config['os'][$os][$key])) {
                return self::$config['os'][$os][$key];
            }

            $os_key = "os.$os.$key";
            if (self::has($os_key)) {
                return self::get($os_key);
            }
        }

        return $default;
    }

    /**
     * Get the merged array from the global and os settings for the specified key.
     * Removes any duplicates.
     * When the arrays have keys, os settings take precedence over global settings
     *
     * @param  string|null  $os  The os name
     * @param  string  $key  period separated config variable name
     * @param  string  $global_prefix  prefix for global setting
     * @param  array  $default  optional array to return if the setting is not set
     * @return array
     */
    public static function getCombined(?string $os, string $key, string $global_prefix = '', array $default = []): array
    {
        $global_key = $global_prefix . $key;

        if (! isset(self::$config['os'][$os][$key])) {
            if (! Str::contains($global_key, '.')) {
                return (array) self::get($global_key, $default);
            }
            if (! self::has("os.$os.$key")) {
                return (array) self::get($global_key, $default);
            }
        }

        if (! self::has("os.$os.$key")) {
            return (array) self::get($global_key, $default);
        }

        return array_unique(array_merge(
            (array) self::get($global_key),
            (array) self::getOsSetting($os, $key)
        ));
    }

    /**
     * Set a variable in the global config
     *
     * @param  mixed  $key  period separated config variable name
     * @param  mixed  $value
     */
    public static function set($key, $value)
    {
        Arr::set(self::$config, $key, $value);
    }

    /**
     * Save setting to persistent storage.
     *
     * @param  mixed  $key  period separated config variable name
     * @param  mixed  $value
     * @return bool if the save was successful
     */
    public static function persist($key, $value)
    {
        try {
            Arr::set(self::$config, $key, $value);
            \App\Models\Config::updateOrCreate(['config_name' => $key], [
                'config_name' => $key,
                'config_value' => $value,
            ]);

            // delete any children (there should not be any unless it is legacy)
            \App\Models\Config::query()->where('config_name', 'like', "$key.%")->delete();

            return true;
        } catch (Exception $e) {
            if (class_exists(Log::class)) {
                Log::error($e);
            }
            if (Debug::isEnabled()) {
                echo $e;
            }

            return false;
        }
    }

    /**
     * Forget a key and all it's descendants from persistent storage.
     * This will effectively set it back to default.
     *
     * @param  string  $key
     * @return int|false
     */
    public static function erase($key)
    {
        self::forget($key);
        try {
            return \App\Models\Config::withChildren($key)->delete();
        } catch (Exception $e) {
            return false;
        }
    }

    /**
     * Check if a setting is set
     *
     * @param  string  $key  period separated config variable name
     * @return bool
     */
    public static function has($key)
    {
        if (isset(self::$config[$key])) {
            return true;
        }

        if (! Str::contains($key, '.')) {
            return false;
        }

        return Arr::has(self::$config, $key);
    }

    /**
     * Serialise the whole configuration to json for use in external processes.
     *
     * @return string
     */
    public static function toJson()
    {
        return json_encode(self::$config);
    }

    /**
     * Get the full configuration array
     *
     * @return array
     */
    public static function getAll()
    {
        return self::$config;
    }

    /**
     * merge the database config with the global config
     * Global config overrides db
     */
    private static function loadDB()
    {
        try {
            \App\Models\Config::get(['config_name', 'config_value'])
                ->each(function ($item) {
                    Arr::set(self::$config, $item->config_name, $item->config_value);
                });
        } catch (QueryException $e) {
            // possibly table config doesn't exist yet
        }

        // load graph types from the database
        self::loadGraphsFromDb(self::$config);
    }

    private static function loadGraphsFromDb(&$config)
    {
        try {
            $graph_types = GraphType::all()->toArray();
        } catch (QueryException $e) {
            // possibly table config doesn't exist yet
            $graph_types = [];
        }

        // load graph types from the database
        foreach ($graph_types as $graph) {
            $g = [];
            foreach ($graph as $k => $v) {
                if (strpos($k, 'graph_') == 0) {
                    // remove leading 'graph_' from column name
                    $key = str_replace('graph_', '', $k);
                } else {
                    $key = $k;
                }
                $g[$key] = $v;
            }

            $config['graph_types'][$g['type']][$g['subtype']] = $g;
        }
    }

    /**
     * Handle defaults that are set programmatically
     */
    private static function processDefaults()
    {
        Arr::set(self::$config, 'log_dir', base_path('logs'));
        Arr::set(self::$config, 'distributed_poller_name', php_uname('n'));

        // set base_url from access URL
        if (isset($_SERVER['SERVER_NAME']) && isset($_SERVER['SERVER_PORT'])) {
            $port = $_SERVER['SERVER_PORT'] != 80 ? ':' . $_SERVER['SERVER_PORT'] : '';
            // handle literal IPv6
            $server = Str::contains($_SERVER['SERVER_NAME'], ':') ? "[{$_SERVER['SERVER_NAME']}]" : $_SERVER['SERVER_NAME'];
            Arr::set(self::$config, 'base_url', "http://$server$port/");
        }

        // graph color copying
        Arr::set(self::$config, 'graph_colours.mega', array_merge(
            (array) Arr::get(self::$config, 'graph_colours.psychedelic', []),
            (array) Arr::get(self::$config, 'graph_colours.manycolours', []),
            (array) Arr::get(self::$config, 'graph_colours.default', []),
            (array) Arr::get(self::$config, 'graph_colours.mixed', [])
        ));
    }

    /**
     * Process the config after it has been loaded.
     * Make sure certain variables have been set properly and
     */
    private static function processConfig()
    {
        // If we're on SSL, let's properly detect it
        if (
            isset($_SERVER['HTTPS']) ||
            (isset($_SERVER['HTTP_X_FORWARDED_PROTO']) && $_SERVER['HTTP_X_FORWARDED_PROTO'] == 'https')
        ) {
            self::set('base_url', preg_replace('/^http:/', 'https:', self::get('base_url', '')));
        }

        self::set('base_url', Str::finish(self::get('base_url', ''), '/'));

        if (! self::get('email_from')) {
            self::set('email_from', '"' . self::get('project_name') . '" <' . self::get('email_user') . '@' . php_uname('n') . '>');
        }

        // Define some variables if they aren't set by user definition in config_definitions.json
        self::set('applied_site_style', self::get('site_style'));
        self::setDefault('html_dir', '%s/html', ['install_dir']);
        self::setDefault('rrd_dir', '%s/rrd', ['install_dir']);
        self::setDefault('mib_dir', '%s/mibs', ['install_dir']);
        self::setDefault('log_dir', '%s/logs', ['install_dir']);
        self::setDefault('log_file', '%s/%s.log', ['log_dir', 'project_id']);
        self::setDefault('plugin_dir', '%s/plugins', ['html_dir']);
        self::setDefault('temp_dir', sys_get_temp_dir() ?: '/tmp');
        self::setDefault('irc_nick', '%s', ['project_name']);
        self::setDefault('irc_chan.0', '##%s', ['project_id']);
        self::setDefault('page_title_suffix', '%s', ['project_name']);
//        self::setDefault('email_from', '"%s" <%s@' . php_uname('n') . '>', ['project_name', 'email_user']);  // FIXME email_from set because alerting config

        // deprecated variables
        self::deprecatedVariable('rrdgraph_real_95th', 'rrdgraph_real_percentile');
        self::deprecatedVariable('fping_options.millisec', 'fping_options.interval');
        self::deprecatedVariable('discovery_modules.cisco-vrf', 'discovery_modules.vrf');
        self::deprecatedVariable('discovery_modules.toner', 'discovery_modules.printer-supplies');
        self::deprecatedVariable('poller_modules.toner', 'poller_modules.printer-supplies');
        self::deprecatedVariable('discovery_modules.cisco-sla', 'discovery_modules.slas');
        self::deprecatedVariable('poller_modules.cisco-sla', 'poller_modules.slas');
        self::deprecatedVariable('oxidized.group', 'oxidized.maps.group');

        // migrate device display
        if (! self::has('device_display_default')) {
            $display_value = '{{ $hostname }}';
            if (self::get('force_hostname_to_sysname')) {
                $display_value = '{{ $sysName }}';
            } elseif (self::get('force_ip_to_sysname')) {
                $display_value = '{{ $sysName_fallback }}';
            }

            self::persist('device_display_default', $display_value);
        }

        // make sure we have full path to binaries in case PATH isn't set
        foreach (['fping', 'fping6', 'snmpgetnext', 'rrdtool', 'traceroute'] as $bin) {
            if (! is_executable(self::get($bin))) {
                self::persist($bin, self::locateBinary($bin));
            }
        }

        if (! self::has('rrdtool_version')) {
            self::persist('rrdtool_version', Rrd::version());
        }
        if (! self::has('snmp.unescape')) {
            self::persist('snmp.unescape', version_compare(Version::get()->netSnmp(), '5.8.0', '<'));
        }
        if (! self::has('reporting.usage')) {
            self::persist('reporting.usage', (bool) Callback::get('enabled'));
        }

        self::populateTime();

        // populate legacy DB credentials, just in case something external uses them.  Maybe remove this later
        self::populateLegacyDbCredentials();
    }

    /**
     * Set default values for defaults that depend on other settings, if they are not already loaded
     *
     * @param  string  $key
     * @param  string  $value  value to set to key or vsprintf() format string for values below
     * @param  array  $format_values  array of keys to send to vsprintf()
     */
    private static function setDefault($key, $value, $format_values = [])
    {
        if (! self::has($key)) {
            if (is_string($value)) {
                $format_values = array_map('\KartsNMS\Config::get', $format_values);
                self::set($key, vsprintf($value, $format_values));
            } else {
                self::set($key, $value);
            }
        }
    }

    /**
     * Copy data from old variables to new ones.
     *
     * @param  string  $old
     * @param  string  $new
     */
    private static function deprecatedVariable($old, $new)
    {
        if (self::has($old)) {
            if (Debug::isEnabled()) {
                echo "Copied deprecated config $old to $new\n";
            }
            self::set($new, self::get($old));
        }
    }

    /**
     * Locate the actual path of a binary
     *
     * @param  string  $binary
     * @return mixed
     */
    public static function locateBinary($binary)
    {
        if (! Str::contains($binary, '/')) {
            $output = `whereis -b $binary`;
            $list = trim(substr($output, strpos($output, ':') + 1));
            $targets = explode(' ', $list);
            foreach ($targets as $target) {
                if (is_executable($target)) {
                    return $target;
                }
            }
        }

        return $binary;
    }

    private static function populateTime()
    {
        $now = time();
        $now -= $now % 300;
        self::set('time.now', $now);
        self::set('time.onehour', $now - 3600); // time() - (1 * 60 * 60);
        self::set('time.fourhour', $now - 14400); // time() - (4 * 60 * 60);
        self::set('time.sixhour', $now - 21600); // time() - (6 * 60 * 60);
        self::set('time.twelvehour', $now - 43200); // time() - (12 * 60 * 60);
        self::set('time.day', $now - 86400); // time() - (24 * 60 * 60);
        self::set('time.twoday', $now - 172800); // time() - (2 * 24 * 60 * 60);
        self::set('time.week', $now - 604800); // time() - (7 * 24 * 60 * 60);
        self::set('time.twoweek', $now - 1209600); // time() - (2 * 7 * 24 * 60 * 60);
        self::set('time.month', $now - 2678400); // time() - (31 * 24 * 60 * 60);
        self::set('time.twomonth', $now - 5356800); // time() - (2 * 31 * 24 * 60 * 60);
        self::set('time.threemonth', $now - 8035200); // time() - (3 * 31 * 24 * 60 * 60);
        self::set('time.sixmonth', $now - 16070400); // time() - (6 * 31 * 24 * 60 * 60);
        self::set('time.year', $now - 31536000); // time() - (365 * 24 * 60 * 60);
        self::set('time.twoyear', $now - 63072000); // time() - (2 * 365 * 24 * 60 * 60);
    }

    public static function populateLegacyDbCredentials()
    {
        $db = config('database.default');

        self::set('db_host', config("database.connections.$db.host", 'localhost'));
        self::set('db_name', config("database.connections.$db.database", 'kartsnms'));
        self::set('db_user', config("database.connections.$db.username", 'kartsnms'));
        self::set('db_pass', config("database.connections.$db.password"));
        self::set('db_port', config("database.connections.$db.port", 3306));
        self::set('db_socket', config("database.connections.$db.unix_socket"));
    }

    /**
     * Check if the config has been loaded yet
     *
     * @return bool
     */
    public static function isLoaded(): bool
    {
        return ! is_null(self::$config);
    }
}
